### PR TITLE
Assignment: Database-Driven Product Server, Part 2

### PR DESCRIPTION
**Pull request by alexhcole95**
## Summary

Completed the second part of the database-driven product server assignment, which was refactoring http handlers.

## Changes

1. Updated README file to include the assignment instructions.
2. Created and implemented proper HTTP handlers in main.go.
3. Modified and tidied directories.

### GET

Screen Shot 2022-11-24 at 7 04 04 PM
Screen Shot 2022-11-24 at 7 09 36 PM

### POST

Screen Shot 2022-11-24 at 7 06 32 PM
Screen Shot 2022-11-24 at 7 06 02 PM

### PUT

Screen Shot 2022-11-24 at 7 07 29 PM
Screen Shot 2022-11-24 at 7 07 15 PM

### DELETE

Screen Shot 2022-11-24 at 7 08 30 PM
Screen Shot 2022-11-24 at 7 08 18 PM

### CODE REVIEW

**Comment by alexhcole95** [comment]
I am having issues with the first GET instruction which is:

`GET /products?limit=<numeric>&sort=<column>`

As you can see from my code (line 59 - 79), I feel like I have an idea of what to do, however it's not working for me. I believe my problem lies with understanding how a request works past the `?limit=<num>&...` part as this is the first time I've worked with that. I've spend a good bit of time on Google and haven't found the appropriate resource yet. Can you point me in the right direction?

File: cmd/database-driven-server/main.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	for rows.Next() {
+		var id int
+		var name string
+		var price float64
+		err = rows.Scan(&id, &name, &price)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		products = append(products, product{ID: id, Name: name, Price: price})
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:53]: No. You don't want to load up an entire database table worth of products into memory. Query the database as needed.

File: cmd/database-driven-server/main.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+	vars := mux.Vars(r)
+	lim, err := strconv.ParseInt(vars["lim"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", 400)
+	}
+	sort := vars["sort"]
+
+	log.Println("Request received! Searching for products...")
+	if lim < 0 || lim > int64(len(products)) {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:70]: No. Remove the use of `products` from your handlers.

File: cmd/database-driven-server/main.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		log.Println("Products not found! Returning status to client...")
+		http.Error(w, "ID not found", 404)
+	} else {
+		log.Println("Products found! Returning products to client...")
+		if _, err = db.Exec("SELECT * FROM products WHERE id = ? LIMIT ?", sort, lim); err != nil {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:75]: `sort` will be assigned to `id` here. Not what you want.

File: cmd/database-driven-server/main.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	log.Println("Receiving request from client...")
+
+	vars := mux.Vars(r)
+	itemID, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", 400)
+	}
+
+	log.Println("Request received! Searching for product...")
+	if itemID < 0 || itemID > int64(len(products)) {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:91]: Remove use of `products` here.

File: cmd/database-driven-server/main.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+		log.Println("Product not found! Returning status to client...")
+		http.Error(w, "ID not found", 404)
+	} else {
+		for _, product := range products {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:95]: No. Query the database for the product you need.

File: cmd/database-driven-server/main.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Unable to read request body", 400)
+	}
+	defer r.Body.Close()
+
+	if err = json.Unmarshal(reqBody, &item); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	items = append(products, item)
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:120]: No use of `products`.

File: cmd/database-driven-server/main.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	products = items
+
+	log.Println("Request received! Adding product to DB...")
+	if item.Name == "" || item.ID == 0 || item.Price == 0 {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:124]: Maybe put all of these validations behind a single method?

File: cmd/database-driven-server/main.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	var updatedProduct product
+	if err = json.Unmarshal(reqBody, &updatedProduct); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	if updatedProduct.Name == "" || updatedProduct.ID == 0 || updatedProduct.Price == 0 {
+		http.Error(w, "Missing required fields", 400)
+	} else {
+		log.Println("Product found! Updating product in DB...")
+		for i, product := range products {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:161]: Remove use of `products`.

File: cmd/database-driven-server/main.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+			if int64(product.ID) == itemID {
+				product.ID = updatedProduct.ID
+				product.Name = updatedProduct.Name
+				product.Price = updatedProduct.Price
+				products[i] = product
+				if _, err = db.Exec("UPDATE products SET name = ?, price = ? WHERE id = ?", updatedProduct.Name, updatedProduct.Price, updatedProduct.ID); err != nil {
+					log.Fatal(err)
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:168]: Don't crash your own server if a query fails.

File: cmd/database-driven-server/main.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+	log.Println("Receiving request from client...")
+	var newProduct []product
+
+	vars := mux.Vars(r)
+	itemID, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", 400)
+	}
+
+	if itemID < 0 || itemID > int64(len(products)) {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:188]: Remove use of `products`.

File: cmd/database-driven-server/main.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+		for _, product := range products {
+			if int64(product.ID) != itemID {
+				newProduct = append(newProduct, product)
+			}
+		}
+		products = newProduct
+
+		log.Println("Product found! Deleting product from DB...")
+		if _, err = db.Exec("DELETE FROM products WHERE id = ?", itemID); err != nil {
+			log.Fatal(err)
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:202]: Don't crash your own server.

**Review by jboursiquot — CHANGES_REQUESTED** [review]
Changes needed.

File: cmd/database-driven-server/main.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+	"encoding/json"
+	"github.com/gorilla/mux"
+	_ "github.com/mattn/go-sqlite3"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+)
+
+var info []product
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:14]: No need to have this as a global.

File: cmd/database-driven-server/main.go [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+	for rows.Next() {
+		var id int
+		var name string
+		var price float64
+		err = rows.Scan(&id, &name, &price)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		info = append(info, product{ID: id, Name: name, Price: price})
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:82]: Just create a local `products` var to hold your products and return that. No need to use the global here.

File: cmd/database-driven-server/main.go [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+	log.Println("Products found! Returning products to client...")
+	if err = json.NewEncoder(w).Encode(info); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func getProductHandler(w http.ResponseWriter, r *http.Request) {
+	log.Println("Receiving request from client...")
+
+	rows, err := db.Query("SELECT * FROM products")
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:94]: Simplify this handler by using the `QueryRow` method instead.

File: cmd/database-driven-server/main.go [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+	if err = json.Unmarshal(reqBody, &item); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	items = append(info, item)
+	info = items
+
+	log.Println("Request received! Adding product to DB...")
+	if item.IsValid() == false {
+		http.Error(w, "Missing required fields", 400)
+	} else {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:153]: No need for an `else` cause here. return early if there's an error.

File: cmd/database-driven-server/main.go [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Unable to read request body", 400)
+	}
+	defer r.Body.Close()
+
+	if err = json.Unmarshal(reqBody, &item); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+	items = append(info, item)
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:147]: Why the need for this `items`?

File: cmd/database-driven-server/main.go [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+
+		log.Println("Product added to DB! Returning status to client...")
+		w.WriteHeader(http.StatusCreated)
+	}
+}
+
+func updateProductHandler(w http.ResponseWriter, r *http.Request) {
+	log.Println("Receiving request from client...")
+
+	rows, err := db.Query("SELECT * FROM products")
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:166]: Simplify with `QueryRow` method.

File: cmd/database-driven-server/main.go [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+		http.Error(w, "Unable to read request body", 400)
+	}
+	defer r.Body.Close()
+
+	var updatedProduct product
+	if err = json.Unmarshal(reqBody, &updatedProduct); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+	if updatedProduct.Name == "" || updatedProduct.ID == 0 || updatedProduct.Price == 0 {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:201]: Put all of this validation behind a `validate() error` method on `Product`.

File: cmd/database-driven-server/main.go [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+		http.Error(w, "Missing required fields", 400)
+	} else {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:203]: No need for `else`. Return early.

File: cmd/database-driven-server/main.go [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+		log.Println("Product found! Updating product in DB...")
+		for i, product := range info {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:205]: You're not going to need to use any loops here if you use `QueryRow`.

File: cmd/database-driven-server/main.go [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+	vars := mux.Vars(r)
+	itemID, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		http.Error(w, "Unable to convert string to int", 400)
+	}
+
+	if itemID < 0 || itemID > int64(len(info)) {
+		log.Println("Product not found! Returning status to client...")
+		http.Error(w, "ID not found", 404)
+	} else {
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:235]: Use early return. No `else` needed.

File: cmd/database-driven-server/main.go [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+		log.Println("Request received! Searching for product...")
+		for _, product := range info {
+			if int64(product.ID) != itemID {
+				newProduct = append(newProduct, product)
+			}
+		}
+		info = newProduct
```
Comment (jboursiquot) [inline cmd/database-driven-server/main.go:242]: Why are you doing this here?

**Review by jboursiquot — CHANGES_REQUESTED** [review]

**Pull request closed without merging**